ipset.go

```diff
@@ -92,34 +105,49 @@ func (s *IPSet) RemovePrefix(p IPPrefix) { s.RemoveRange(p.Range()) }
 
 // RemoveRange removes r from s.
 func (s *IPSet) RemoveRange(r IPRange) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.removeRangeLocked(r)
+}
+
+func (s *IPSet) removeRangeLocked(r IPRange) {
 	if r.Valid() {
 		s.out = append(s.out, r)
+		s.normalized = false
 	}
 }
 
 // AddSet adds all ranges in b to s.
 func (s *IPSet) AddSet(b *IPSet) {
-	for _, r := range b.Ranges() {
-		s.AddRange(r)
+	rr := b.Ranges()
```

Comment:
Why not rangesLock and moved mutex higher? I’d expect any given method to be atomic.

Reply:
Lock ordering. If you want AddSet to be completely atomic, we'd have to simultaneously lock s and b, and that feels really easy to misuse in the caller (e.g. two goroutines, one calling a.AddSet(b) and another calling b.AddSet(a) could deadlock).

Given that, I chose to go with the semantics that _some_ consistent snapshot of b will be used, but not make promises beyond that.